hash-parser/src/emit.rs

```diff
@@ -0,0 +1,191 @@
+use std::iter;
+
+use crate::{ast, grammar::Rule, location::Location, modules::ModuleIdx};
+
+struct AstBuilder {
+    pos: Location,
+    module: ModuleIdx,
+}
+
+impl AstBuilder {
+    pub fn from_pair(pair: &pest::iterators::Pair<'_, Rule>) -> AstBuilder {
+        let span = pair.as_span();
+        let pos = Location::Span(span.start(), span.end());
+        AstBuilder {
+            pos,
+            module: 0, // @Todo
+        }
+    }
+
+    pub fn from_node<T>(node: &ast::AstNode<T>) -> AstBuilder {
+        AstBuilder {
+            pos: node.pos,
+            module: node.module,
+        }
+    }
+
+    pub fn node<T>(&self, inner: T) -> ast::AstNode<T> {
+        ast::AstNode {
+            body: Box::new(inner),
+            pos: self.pos,
+            module: self.module,
+        }
+    }
+}
+
+type HashPair<'a> = pest::iterators::Pair<'a, Rule>;
+
+impl ast::IntoAstNode<ast::Name> for HashPair<'_> {
+    fn into_ast(self) -> ast::AstNode<ast::Name> {
+        match self.as_rule() {
+            Rule::name => AstBuilder::from_pair(&self).node(ast::Name {
+                string: self.as_str().to_owned(),
+            }),
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl ast::IntoAstNode<ast::AccessName> for HashPair<'_> {
+    fn into_ast(self) -> ast::AstNode<ast::AccessName> {
+        match self.as_rule() {
+            Rule::access_name => AstBuilder::from_pair(&self).node(ast::AccessName {
+                names: self
+                    .into_inner()
+                    .map(|p| p.into_ast())
+                    .collect(),
+            }),
+            _ => unreachable!(),
+        }
+    }
+}
+
+const FUNCTION_TYPE_NAME: &str = "Function";
+const TUPLE_TYPE_NAME: &str = "Tuple";
+
+fn single_access_name(ab: &AstBuilder, name: &str) -> ast::AstNode<ast::AccessName> {
+    ab.node(ast::AccessName {
+        names: vec![ab.node(ast::Name {
+            string: name.to_owned(),
+        })],
+    })
+}
+
+impl ast::IntoAstNode<ast::Type> for HashPair<'_> {
+    fn into_ast(self) -> ast::AstNode<ast::Type> {
```

Comment:
I guess we will need to add implementations for each `ast::*` ?

Reply:
Yes indeed.